### PR TITLE
Issue 2578: enable TLS for REST interface

### LINKED ISSUE #2578: Pravega security: Enable TLS for rest server
**Problem description**
The REST server does not enable encrypted communication. There should be an option of enabling TLS for rest.

**Problem location**
`RESTServer.java` for controller.

**Suggestions for an improvement**
Implement TLS for REST APIs and support configuration of it through controller properties.

### PR DESCRIPTION
**Pull request by arvindkandhare**
**Change log description**
Enables TLS for REST interface.
This is part of #1669 .
As part of enabling encryption on all the channels, the REST API also should have functionality where TLS can be enabled. The TLS for REST API can be enabled using `RESTServerConfig`. The TLS keyfile and password file to access the key file can be speicified using these configurations.
These files follow `JKS` format.
 
**Purpose of the change**
This fixes #2578.

**What the code does**
1. Introduces config related to TLS configuration for the REST API
2. Implements a TLS enabled server based on these settings
3. Test cases for the REST API.

**How to verify it**
The unit tests pass. Controller is available on `https://` URL when TLS is enabled.

### CODE REVIEW

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 1]
```diff
@@ -68,7 +70,17 @@ protected void startUp() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "startUp");
         try {
             log.info("Starting REST server listening on port: {}", this.restServerConfig.getPort());
-            httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true);
+            if (restServerConfig.isEnableTls()) {
+                SSLContextConfigurator sslCon = new SSLContextConfigurator();
+                sslCon.setKeyStoreFile(restServerConfig.getKeyFilePath());
+                sslCon.setKeyStorePass("1111_aaaa");
```
Comment (EronWright) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: (hardcoded password)

File: config/standalone-config.properties [hunk 2]
```diff
@@ -23,22 +23,22 @@
...
-#singlenode.passwd=1111_aaaa
+singlenode.passwd=1111_aaaa
 
 # Enable TLS for segmentstore and controller.
-#singlenode.enableTls=true
+singlenode.enableTls=true
 
 # Enable authentication/authorization.
-#singlenode.enableAuth=true
+singlenode.enableAuth=true
```
Comment (EronWright) [inline config/standalone-config.properties]: I don't think TLS or auth should be enabled by default

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java [hunk 3]
```diff
@@ -63,8 +69,57 @@ public void tearDown() {
 
     @Test
     public void test() {
-        String streamResourceURI = "http://localhost:" + serverConfig.getPort() + "/ping";
+        String streamResourceURI = getURLScheme() + "://localhost:" + serverConfig.getPort() + "/ping";
```
Comment (EronWright) [inline controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: Could we maybe use `javax.ws.rs.core.UriBuilder` or something?

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java [hunk 4]
```diff
@@ -63,8 +69,57 @@ public void tearDown() {
...
+            return "http";
+        }
+    }
+
+    public static class SecurePingTest extends PingTest {
+
+        @Override
+        protected Client createJerseyClient() {
+            SslConfigurator sslConfig = SslConfigurator.newInstance()
+                                                       .trustStoreFile("../config/bookie.truststore.jks");
```
Comment (EronWright) [inline controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: It just seems bad to hardcode these paths into a test because it presupposed the working directory.   I find that Gradle and IntelliJ have different defaults in that respect.  I think the ideal would be for these files to be in `src/test/resources` and extracted as temp files.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java [hunk 5]
```diff
@@ -119,6 +119,7 @@ private void bootstrapRequestWriters() {
                                     ClientConfig.builder().controllerURI(configuration.getControllerUri())
                                                 .credentials(new DefaultCredentials(configuration.getAuthPassword(), configuration.getAuthUsername()))
                                                 .trustStore(configuration.getTlsCertFile())
+                                                .validateHostName(false)
```
Comment (EronWright) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java]: To my knowledge the only legit scenario for disabling hostname verification is when the truststore is specially crafted to allow ONLY the server's TLS certificate.   Please be extra careful about this.

Comment (arvindkandhare) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java]: Will take care of this in #2582.

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java (hunk 4 above)
Comment (arvindkandhare) [reply controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: We need these files for easy running of `standalone`. Reusing these here.

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java (hunk 3 above)
Comment (arvindkandhare) [reply controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: Using `UriBuilder`. Fixed.

File: config/standalone-config.properties (hunk 2 above)
Comment (arvindkandhare) [reply config/standalone-config.properties]: Fixed.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 1 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed. Reading from a config.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 6]
```diff
@@ -68,12 +74,39 @@ protected void startUp() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "startUp");
         try {
             log.info("Starting REST server listening on port: {}", this.restServerConfig.getPort());
-            httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true);
+            if (restServerConfig.isEnableTls()) {
+                SSLContextConfigurator sslCon = new SSLContextConfigurator();
```
Comment (shiveshr) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: dont we need to specify server auth certificate somewhere for tls.. is the keystorefile the certificate itself?

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java (hunk 5 above)
Comment (shiveshr) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java]: This class doesnt make use of the REST interface to communicate with controller. 
why is this change relevant here?

Comment (arvindkandhare) [reply segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java]: This is a temp fix for #2582. Will merge that PR into this before merging.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 6 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: That is correct. keystorefile is the certificate store in `JKS` format.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 7]
```diff
@@ -68,12 +74,39 @@ protected void startUp() {
...
+                sslCon.setKeyStoreFile(restServerConfig.getKeyFilePath());
+                sslCon.setKeyStorePass(loadPasswdFromFile(restServerConfig.getKeyFilePasswordPath()));
+                httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true,
+                        new SSLEngineConfigurator(sslCon, false, false, false));
+            } else {
+                httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true);
+            }
+        } catch (Exception e) {
+          log.error("Exception while creating HTTP server", e);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: This exception needs to be rethrown. Otherwise the `AbstractIdleService` will have started up normally but there's no server to back it up.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 8]
```diff
@@ -68,12 +74,39 @@ protected void startUp() {
...
         } finally {
             LoggerHelpers.traceLeave(log, this.objectId, "startUp", traceId);
         }
     }
 
+    private String loadPasswdFromFile(String keyFilePasswordPath) {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Please don't abbreviate. "passwd" -> "password"

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java [hunk 9]
```diff
@@ -26,4 +26,22 @@
      * @return The port on which controller gRPC listens.
      */
     int getPort();
+
+    /**
+     * Flag which denotes whether TLS is enabled.
+     * @return Flag which denotes whether TLS is enabled.
+     */
+    boolean isEnableTls();
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java]: does `isTlsEnabled` sound better?

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java [hunk 10]
```diff
@@ -26,4 +26,22 @@
...
+
+    /**
+     * File which contains the key file for the TLS connection.
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java]: This returns a Path, not a file.

Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java]: Fixed.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java (hunk 9 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServerConfig.java]: Fixed.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 8 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 7 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed. Dropping the `catch` block so that the earlier behavior is maintained.

File: standalone/src/main/java/io/pravega/local/InProcPravegaCluster.java [hunk 11]
```diff
@@ -338,6 +338,9 @@ private ControllerServiceMain startLocalController(int controllerId) {
         RESTServerConfig restServerConfig = RESTServerConfigImpl.builder()
                 .host("0.0.0.0")
                 .port(this.restServerPort)
+                .tlsEnabled(this.enableTls)
+                .keyFilePath("../config/bookie.keystore.jks")
+                .keyFilePasswordPath("../config/bookie.keystore.jks.passwd")
```
Comment (EronWright) [inline standalone/src/main/java/io/pravega/local/InProcPravegaCluster.java]: Please don't apply default paths to the in-proc server!

**Comment by EronWright** [comment]
Please stop breaking the emulator with hardcoded configuration paths.

File: standalone/src/main/java/io/pravega/local/InProcPravegaCluster.java (hunk 11 above)
Comment (arvindkandhare) [reply standalone/src/main/java/io/pravega/local/InProcPravegaCluster.java]: Fixed.

File: standalone/src/main/java/io/pravega/local/SingleNodeConfig.java [hunk 12]
```diff
@@ -28,6 +28,10 @@
     public final static Property<Boolean> ENABLE_TLS = Property.named("enableTls", false);
     public final static Property<Boolean> ENABLE_AUTH = Property.named("enableAuth", false);
     public final static String PROPERTY_FILE = "singlenode.configurationFile";
+    public final static Property<String> KEYSTORE_JKS = Property.named("singlenode.keyStoreJKS",
+            "../config/bookie.keystore.jks");
+    public final static Property<String> KEYSTORE_JKS_PASSWORD = Property.named("singlenode.keyStoreJKSPassowrd",
```
[the remark this answers is not in the document]
Comment (arvindkandhare) [reply standalone/src/main/java/io/pravega/local/SingleNodeConfig.java]: These are just the defaults. Can be overridden by specifying the values in `singlenode_config.properties`.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 13]
```diff
@@ -68,12 +74,37 @@ protected void startUp() {
         long traceId = LoggerHelpers.traceEnterWithContext(log, this.objectId, "startUp");
         try {
             log.info("Starting REST server listening on port: {}", this.restServerConfig.getPort());
-            httpServer = GrizzlyHttpServerFactory.createHttpServer(baseUri, resourceConfig, true);
+            if (restServerConfig.isTlsEnabled()) {
+                SSLContextConfigurator sslCon = new SSLContextConfigurator();
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: `Con` is supposed to be context? Can we be more descriptive, please?

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 14]
```diff
@@ -68,12 +74,37 @@ protected void startUp() {
...
+            }
         } finally {
             LoggerHelpers.traceLeave(log, this.objectId, "startUp", traceId);
         }
     }
 
+    private String loadPasswordFromFile(String keyFilePasswordPath) {
+
+        if (Strings.isNullOrEmpty(keyFilePasswordPath)) {
+            return "";
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Is it correct to return empty string? Shouldn't it the be an error? I'd like to understand the consequence of returning an empty string here.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 15]
```diff
@@ -68,12 +74,37 @@ protected void startUp() {
...
+        }
+        File passwdFile = new File(keyFilePasswordPath);
+        if (passwdFile.length() == 0) {
+            return "";
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Same question here about returning an empty string.

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java (hunk 4 above)
Comment (fpj) [reply controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: I feel that @arvindkandhare has not addressed the comment. The key to @EronWright 's comment is:

>  it presupposed the working directory. I find that Gradle and IntelliJ have different defaults in that respect.

My recommendation is that you create a temporary file for this test, populate it, use it in the test, and drop it.

File: controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java [hunk 16]
```diff
@@ -63,8 +71,58 @@ public void tearDown() {
...
+            return ClientBuilder.newBuilder().sslContext(sslContext)
+                                .hostnameVerifier((s1, s2) -> true)
+                                .build();
+        }
+
+        @Override
+        RESTServerConfig getServerConfig() {
+            return RESTServerConfigImpl.builder().host("localhost").port(TestUtils.getAvailableListenPort())
+                                       .tlsEnabled(true)
+                                       .keyFilePath("../config/bookie.keystore.jks")
```
Comment (fpj) [inline controller/src/test/java/io/pravega/controller/rest/v1/PingTest.java]: More hardcoded paths.

File: segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java [hunk 17]
```diff
@@ -117,6 +117,7 @@ private void bootstrapRequestWriters() {
                             factory = ClientFactory.withScope(NameUtils.INTERNAL_SCOPE_NAME,
                                     ClientConfig.builder().controllerURI(configuration.getControllerUri())
                                                 .trustStore(configuration.getTlsCertFile())
+                                                .validateHostName(false)
```
Comment (fpj) [inline segmentstore/server/host/src/main/java/io/pravega/segmentstore/server/host/stat/AutoScaleProcessor.java]: This change seems to be out of scope.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 13 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed. Renamed it to be more descriptive.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 14 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Empty string means the password is not used. We return empty string if user have not specified the password path. Which means that user is not planning to use a password for the keyfile. Added a comment to the code.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 15 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: The comment in the code covers this path too.

**Comment by EronWright** [comment]
I don't know if it has been addressed yet, the issue of the Standalone mode vs emulator mode.  Only in the former mode is it legitimate to have default paths.

**Comment by arvindkandhare** [comment]
[Enable encryption for REST interface to controller](https://trello.com/c/bv0cdWv5/26-enable-encryption-for-rest-interface-to-controller)

**Comment by arvindkandhare** [comment]
[Enable encryption for REST interface to controller](https://trello.com/c/bv0cdWv5/26-enable-encryption-for-rest-interface-to-controller)

**Comment by arvindkandhare** [comment]
[Enable encryption for REST interface to controller](https://trello.com/c/bv0cdWv5/26-enable-encryption-for-rest-interface-to-controller)

File: build.gradle [hunk 18]
```diff
@@ -441,6 +441,7 @@ project('shared:controller-api') {
 project('controller') {
     sourceSets {
         main.resources.srcDirs += "$projectDir/src/conf"
+        test.resources.srcDirs += "$rootDir/config"
```
Comment (EronWright) [inline build.gradle]: I think we should decide on the purpose of the `config` directory, because it is apparently playing quite a few roles here.    There's a variety of scenarios:
1. the default configuration file included in the Pravega distribution, for cluster mode.
2. the default configuration file for the standalone mode.
3. the configuration file (and associated data files) for integration tests.

The change seen above (line 444) is obviously related to (3) but it would probably be best to use a proper test resource directory.  Meanwhile, it is questionable that we provide a jks or passwd file.

File: standalone/src/main/java/io/pravega/local/SingleNodeConfig.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (EronWright) [reply standalone/src/main/java/io/pravega/local/SingleNodeConfig.java]: Ideally one would not need these defaults.

File: build.gradle (hunk 18 above)
Comment (arvindkandhare) [reply build.gradle]: This is correct. About 1 and 2 cluster mode and standalone mode is just two different ways of distributing the same application. They just refer different files in the same `config` directory. Do you recommend finding different locations for configurations of different components?

> Meanwhile, it is questionable that we provide a jks or passwd file.

We have discussed this earlier. This is for user convenience. Admin can create and deploy his own version once he tries the default ones out.

P.S. Both the discussions are not strictly related to the context of the PR :). If we agree this is the way to go, I will log issues for these.

File: standalone/src/main/java/io/pravega/local/SingleNodeConfig.java (hunk 12 above)
[the remark this answers is not in the document]
Comment (arvindkandhare) [reply standalone/src/main/java/io/pravega/local/SingleNodeConfig.java]: Yes. Dropping the defaults. They would have covered the case where the user has messed with singlenode config. But this is a real corner scenario. Either way works.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 19]
```diff
@@ -68,12 +74,39 @@ protected void startUp() {
...
+        }
+        try {
+            return new String(FileUtils.readFileToByteArray(passwdFile)).trim();
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: You need to specify the encoding, otherwise this will try to decode the file using the default charset, which may not match what's in the file.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 20]
```diff
@@ -68,12 +75,39 @@ protected void startUp() {
...
     }
 
+    private String loadPasswordFromFile(String keyFilePasswordPath) {
+        // In case the path is not specified, return empty string. This means the password is not used.
+        // In case of error, return empty password. Which will fail the SSL connection if the password is expected.
+
+        if (Strings.isNullOrEmpty(keyFilePasswordPath)) {
+            return "";
+        }
+        File passwdFile = new File(keyFilePasswordPath);
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: please be consistent in your naming. Either choose "password" or "passwd". I prefer the former.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 21]
```diff
@@ -68,12 +75,39 @@ protected void startUp() {
...
+        if (passwdFile.length() == 0) {
+            return "";
+        }
+        try {
+            return new String(FileUtils.readFileToByteArray(passwdFile), Charsets.UTF_8).trim();
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: What happens if the file is too large? This can easily lead to OOM exceptions and could possibly be used as a way to crash the process. Please have a check to validate that the file isn't too big. Figure out a reasonable maximum length (4KB?) and fail if the file is too big.

File: controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java [hunk 22]
```diff
@@ -21,13 +21,19 @@
 public class RESTServerConfigImpl implements RESTServerConfig {
     private final String host;
     private final int port;
+    private final boolean tlsEnabled;
+    private final String keyFilePath;
+    private final String keyFilePasswordPath;
 
     @Builder
-    RESTServerConfigImpl(final String host, final int port) {
+    RESTServerConfigImpl(final String host, final int port, boolean tlsEnabled, String keyFilePath, String keyFilePasswordPath) {
```
Comment (andreipaduroiu) [inline controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java]: Should keyFilePath or keyPasswordPath be non-null, non-empty? Consider validating them.

Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java]: They need not be non-null, non-empty. In case of `tlsEnabled` false, they are not used at all.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 21 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed. Added a check on the filesize. Currently it is 4MB.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 20 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Fixed. Using `password`.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java [hunk 23]
```diff
@@ -68,12 +77,43 @@ protected void startUp() {
...
+
+        String password = "";
+        if (!Strings.isNullOrEmpty(keyFilePasswordPath)) {
+            File passwordFile = new File(keyFilePasswordPath);
+            if (passwordFile.length() != 0) {
+                if (passwordFile.length() > MAX_PASSWORD_LENGTH) {
+                    log.warn("Very large password file. Not parsing.");
+                } else {
+                    try {
+                        password = new String(FileUtils.readFileToByteArray(passwordFile), Charsets.UTF_8).trim();
```
Comment (fpj) [inline controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Why are we not propagating the exception? Shouldn't the application be made aware explicitly that there is a problem with the password file rather than proceed (almost) silently. The warn message is good, but this is important enough that the application needs to be warned explicitly.

Also, I assume that if we require a password, but we can't get it, then the server will fail because it won't be able to open the key file, yes? How's it going to fail in this case? In particular, what happens to this call:

https://github.com/pravega/pravega/pull/2587/files#diff-120984f086091c8ab4ed8a3cb8ed6cefR83

File: controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java (hunk 22 above)
Comment (fpj) [reply controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java]: I think @andreipaduroiu is requesting that you add more pre-conditions here.

File: build.gradle (hunk 18 above)
Comment (arvindkandhare) [reply build.gradle]: Created issue https://github.com/pravega/pravega/issues/2984 to track this as this is not only limited to this test but is applicable for all the integration/unit tests.

File: controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java (hunk 22 above)
Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/impl/RESTServerConfigImpl.java]: @fpj, thanks for clarifying. Added more preconditions.

File: controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java (hunk 23 above)
Comment (fpj) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Thinking about this again, I think the logic is that password is going to be empty and the SSL connection will fail because it expects one. I'd like to see a test case that covers this scenario, showing that the user can't bypass our security when using an invalid password file. Is large file the only way this can fail?

Comment (arvindkandhare) [reply controller/src/main/java/io/pravega/controller/server/rest/RESTServer.java]: Added a test case which validates that grizzly server does not start when the password does not match.
